### PR TITLE
Solve refacto

### PR DESCRIPTION
**Pull request by Breush**
- `Method::CRA` nows really holds the iteration method and also a dispatch type (whether to use MPI) ;
- broke solve solutions into different files according to the method used ;
- renamed some CRA structures to be more consistent (CraBuilder etc.) ;
- replaced specifier's matrix shape by a flag (and removed extra symmetric info) ;
- `reinterpret_cast<>` are used instead of copying the methods ;
- `make check`, `make benchmarks`, `make examples` are all right
- `make fullcheck` is strictly equivalent to master's

### CODE REVIEW

File: linbox/algorithms/rational-solver.inl [hunk 1]
```diff
@@ -1375,12 +1379,14 @@ namespace LinBox
...
 		}
+
+            std::cout << "ouch" << std::endl;
 		return SS_FAILED; //all primes were bad
 	}
 
 
 
-
+#if 0
```
Comment (ClementPernet) [inline linbox/algorithms/rational-solver.inl]: restore this block and test it. If test fail, comment it out

File: linbox/algorithms/wiedemann.inl [hunk 2]
```diff
@@ -53,568 +53,493 @@
...
+        Singularity singularity = _traits.singularity;
+        if (A.rowdim() != A.coldim()) _traits.singularity = (singularity = Singularity::Singular);
+        ReturnStatus status = FAILED;
+
+        auto tries = _traits.trialsBeforeFailure;
+
+        size_t r = (size_t)-1;
+
+        // Rank == 0 means unknown
+        if (_traits.rank != 0) r = _traits.rank;
```
Comment (ClementPernet) [inline linbox/algorithms/wiedemann.inl]: revert to enum with value -1 for UNKNOWN

File: linbox/algorithms/wiedemann.inl [hunk 3]
```diff
@@ -53,568 +53,493 @@
...
+
+        Polynomial m_A(A.field());
+        Vector z(A.field());
+        bool ret = true;
+
+        {
+            // Make it just Blackbox trait and not wiedemann:
+            // Might need extension field for minpoly
+            // Might also also use better method than Wiedemann ...
+            // @fixme Just reinterpret_cast<const Blackbox&>(_traits) when minpoly uses Method
```
Comment (ClementPernet) [inline linbox/algorithms/wiedemann.inl]: fixme

File: linbox/blackbox/ntl-hankel.inl [hunk 4]
```diff
@@ -219,7 +219,7 @@ namespace LinBox
 	void Hankel<Field>::setToUniModLT()
 	{
 		int L = int(this->rowDim-1)<<1;
-		this->shape.shape(BlackboxSpecifier::UNIMOD_LT);
+		this->shape.shape(Method::Blackbox::UNIMOD_LT);
```
Comment (ClementPernet) [inline linbox/blackbox/ntl-hankel.inl]: fix it to latest method names (install ntl)

File: linbox/blackbox/toeplitz.inl [hunk 5]
```diff
@@ -83,7 +83,7 @@ namespace LinBox
 		sysDim =               // Default dimension is 0
 		rowDim =               // Default row dim is 0
 		this->colDim = 0;            // Default col dim is 0
-		shape.shape(BlackboxSpecifier::TOEPLITZ);
+		shape.shape(Method::Blackbox::TOEPLITZ);
```
Comment (ClementPernet) [inline linbox/blackbox/toeplitz.inl]: fix it

File: linbox/blackbox/toeplitz.inl [hunk 6]
```diff
@@ -97,7 +97,7 @@ namespace LinBox
 		P(PF), field_(&(PF.getCoeffField()))
 	{
 		sysDim = rowDim = this->colDim = 0;
-		shape.shape(BlackboxSpecifier::TOEPLITZ);
+		shape.shapeFlags &= Shape::Toeplitz;
```
Comment (ClementPernet) [inline linbox/blackbox/toeplitz.inl]: fix it

File: linbox/blackbox/toeplitz.inl [hunk 7]
```diff
@@ -111,7 +111,8 @@ namespace LinBox
 						     , size_t n ) :
 		P(PF), field_(&(PF.getCoeffField())), rowDim(m), colDim(n), pdata(p)
 	{
-		shape.shape(BlackboxSpecifier::TOEPLITZ);
+		shape.shapeFlags &= Shape::Toeplitz;
```
Comment (ClementPernet) [inline linbox/blackbox/toeplitz.inl]: fix it

File: linbox/blackbox/toeplitz.inl [hunk 8]
```diff
@@ -63,7 +63,7 @@ namespace LinBox
 		sysDim =               // Default dimension is 0
 		rowDim =               // Default row dim is 0
 		this->colDim = 0;            // Default col dim is 0
-		shape.shape(BlackboxSpecifier::TOEPLITZ);
+		shape.shapeFlags &= Shape::Toeplitz;
```
Comment (ClementPernet) [inline linbox/blackbox/toeplitz.inl]: fix it

File: linbox/solutions/charpoly.h [hunk 9]
```diff
@@ -376,9 +375,9 @@ namespace LinBox
 		size_t n=A.coldim();
 		double hadamarcp = n/2.0*(log(double(n))+2*log(double(max))+0.21163275)/log(2.0);
 
-		ChineseRemainder< FullMultipCRA<Field > > cra(hadamarcp);
+		Cra< CraBuilderFullMultip<Field > > cra(hadamarcp);
 #endif
-		ChineseRemainder< EarlyMultipCRA<Field > > cra(3UL);
+		Cra< CraBuilderEarlyMultip<Field > > cra(3UL);
```
Comment (ClementPernet) [inline linbox/solutions/charpoly.h]: idem

File: linbox/solutions/solve/solve-cra.h [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+        PrimeIterator<LinBox::IteratorCategories::HeuristicTag> primeGenerator(bits);
+        CraRebinderSolver<Matrix, Vector, IterationMethod> iteration(A, b, m.iterationMethod);
+
+        // @note The result is stored to Integers, and will be converted
+        // later back.
+        BlasVector<Givaro::ZRing<Integer>> num(F, A.coldim());
+        Integer den(1);
+
+        auto hb = RationalSolveHadamardBound(A, b);
+        double hadamardLogBound = 1.0 + hb.numLogBound + hb.denLogBound; // = ln2(2 * N * D)
```
Comment (ClementPernet) [inline linbox/solutions/solve/solve-cra.h]: this should be returned by a  dedicated method `solutionLogBound` of `RationalSolveHadamardBound`

**Comment by ClementPernet** [comment]
Please add a treatment for `RationalTag`. This is tested in `examples/solverat.C` and `examples/sparsesolverat.C`.

File: linbox/solutions/solve/solve-auto.h [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+                               const Method::Auto& m)
+    {
+        return solveInPlace(x, A, b, tag, reinterpret_cast<const Method::Dixon&>(m));
+    }
+
+    /**
+     * \brief Solve in place specialisation for Auto with DenseMatrix and non-IntegerTag.
+     */
+    template <class ResultVector, class Field, class Vector, class CategoryTag>
+    typename std::enable_if<!std::is_same<CategoryTag, RingCategories::IntegerTag>::value, ResultVector&>::type solveInPlace(
```
Comment (ClementPernet) [inline linbox/solutions/solve/solve-auto.h]: When `RationalTag` is supported, change this enable_if switch

File: linbox/solutions/solve/solve-auto.h [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+        ResultVector& x, DenseMatrix<Field>& A, const Vector& b, const CategoryTag& tag, const Method::Auto& m)
+    {
+        return solveInPlace(x, A, b, tag, reinterpret_cast<const Method::DenseElimination&>(m));
+    }
+
+    /**
+     * \brief Solve in place specialisation for Auto with SparseMatrix and non-IntegerTag.
+     */
+    template <class ResultVector, class... MatrixArgs, class Vector, class CategoryTag>
+    typename std::enable_if<!std::is_same<CategoryTag, RingCategories::IntegerTag>::value, ResultVector&>::type solveInPlace(
```
Comment (ClementPernet) [inline linbox/solutions/solve/solve-auto.h]: same here

File: linbox/solutions/solve.h [hunk 13]
```diff
@@ -20,1060 +18,317 @@
...
+     *      - Otherwise  > Error
+     * - Method::SymbolicNumericOverlap
+     *      - IntegerTag
+     *      |   - DenseMatrix > `RationalSolverSN<Ring, LPS<FMatrix>>`
+     *      |   - Otherwise   > Error
+     *      - Otherwise  > Error
+     * - Method::SymbolicNumericNorm
+     *      - IntegerTag > `RationalSolver<..., Method::SymbolicNumericNorm>`
+     *      - Otherwise  > Error
+     *
```
Comment (ClementPernet) [inline linbox/solutions/solve.h]: mention RationalTag

File: linbox/solutions/solve.h [hunk 14]
```diff
@@ -20,1060 +18,317 @@
...
-                     const Method::DenseElimination& m)
+    /**
+     * \brief Solve specialisation on IntegerTag with Vector<QField> as result.
+     *
+     * This forward to the rational interface (num, den).
+     * But will only work if the ResultVector if a vector of some Rational type.
+     */
+    template <class ResultVector, class Matrix, class Vector, class SolveMethod>
+    typename std::enable_if<std::is_same<typename SolveMethod::CategoryTag, RingCategories::IntegerTag>::value,
+                            ResultVector&>::type
```
Comment (ClementPernet) [inline linbox/solutions/solve.h]: also if RationalTag

**Comment by Breush** [comment]
@ClementPernet Ok, I fixed the tests, everything is fine for me.

**Comment by ClementPernet** [comment]
LGTM, MERRRGIIINNNG!

**Pull request merged**